File: 125-valid-palindrome/cpp/main.cpp
class Solution {
 public:
  bool isPalindrome(string s) {
    int l = 0;
    int r = s.length() - 1;

    while (l <= r) {
      while (l < s.length() && !isAlphaNum(s[l])) {
        ++l;
      }

      while (r >= 0 && !isAlphaNum(s[r])) {
        --r;
      }

      if (l > r) {
        return true;
      }

      if (toLower(s[l]) != toLower(s[r])) {
        return false;
      }

      ++l;
      --r;
    }

    return true;
  }

  bool isAlphaNum(char c) {
    return (c >= '0' && c <= '9') ||
           ((toLower(c) >= 'a') && (toLower(c) <= 'z'));
  }

  char toLower(char c) {
    if (c >= 'A' && c <= 'Z') {
      return c + 'a' - 'A';
    }
    return c;
  }
};

File: 125-valid-palindrome/go/main.go
package main

func isPalindrome(s string) bool {
	l := 0
	r := len(s) - 1

	for l <= r {
		for (l < len(s)-1) && !isAlphaNum(rune(s[l])) {
			l++
		}

		for (r >= 0) && !isAlphaNum(rune(s[r])) {
			r--
		}

		if l < r && toLower(rune(s[l])) != toLower(rune(s[r])) {
			return false
		}

		l++
		r--
	}

	return true
}

func isAlphaNum(c rune) bool {
	if (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') {
		return true
	}
	return false
}

func toLower(c rune) rune {
	if c >= 'A' && c <= 'Z' {
		return c + 'a' - 'A'
	}
	return c
}
